# PIT PD Vasicek

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings('ignore')

In [2]:
df=pd.read_excel("1_TTC_PD.xlsx",sheet_name="Vasicek Model")

In [3]:
df

observation_date Performance   Annual DR       HPI     UR
0        2003-10-01   2004-01-01   0.036923  0.010458  0.057
1        2004-01-01   2004-04-01   0.035484  0.010575  0.056
2        2004-04-01   2004-07-01   0.032573  0.009930  0.055
3        2004-07-01   2004-10-01   0.029851  0.010189  0.055
4        2004-10-01   2005-01-01   0.028481  0.012283  0.053
..              ...          ...        ...       ...    ...
73       2022-01-01   2022-04-01   0.024169  0.013562  0.037
74       2022-04-01   2022-07-01   0.022222 -0.003378  0.035
75       2022-07-01   2022-10-01   0.024465 -0.001722  0.036
76       2022-10-01   2023-01-01   0.022923 -0.000484  0.034
77       2023-01-01   2023-04-01   0.026012  0.004668  0.034

[78 rows x 5 columns]

In [13]:
df.shape

(78, 5)

1) Taking norm.inv on both the side for the vasicek formula.
2) we find Actual PIT Score:- norm.inv(Annual DR's), taking seed values of Z range and rho


In [7]:
from scipy.stats import norm

In [9]:
df['actual_PIT_score']=norm.ppf(df['Annual DR'])

In [11]:
df

observation_date Performance   Annual DR       HPI     UR  actual_PIT_score
0        2003-10-01   2004-01-01   0.036923  0.010458  0.057         -1.787565
1        2004-01-01   2004-04-01   0.035484  0.010575  0.056         -1.805684
2        2004-04-01   2004-07-01   0.032573  0.009930  0.055         -1.844251
3        2004-07-01   2004-10-01   0.029851  0.010189  0.055         -1.882992
4        2004-10-01   2005-01-01   0.028481  0.012283  0.053         -1.903602
..              ...          ...        ...       ...    ...               ...
73       2022-01-01   2022-04-01   0.024169  0.013562  0.037         -1.974382
74       2022-04-01   2022-07-01   0.022222 -0.003378  0.035         -2.009875
75       2022-07-01   2022-10-01   0.024465 -0.001722  0.036         -1.969204
76       2022-10-01   2023-01-01   0.022923 -0.000484  0.034         -1.996815
77       2023-01-01   2023-04-01   0.026012  0.004668  0.034         -1.942942

[78 rows x 6 columns]

For Fitted PIT Score:-

In [14]:
rho=0.01 #seed value

In [16]:
TTC_PD=0.0524

In [18]:
threshold= norm.ppf(TTC_PD)
threshold

-1.622015637145923

In [22]:
# Z range seed value:-
np.random.seed(42)
df['z']=np.random.normal(0,1,len(df))    #mean--> 0 , std dev--> 1 and no of values--> len(df)

In [24]:
df

observation_date Performance   Annual DR       HPI     UR  \
0        2003-10-01   2004-01-01   0.036923  0.010458  0.057   
1        2004-01-01   2004-04-01   0.035484  0.010575  0.056   
2        2004-04-01   2004-07-01   0.032573  0.009930  0.055   
3        2004-07-01   2004-10-01   0.029851  0.010189  0.055   
4        2004-10-01   2005-01-01   0.028481  0.012283  0.053   
..              ...          ...        ...       ...    ...   
73       2022-01-01   2022-04-01   0.024169  0.013562  0.037   
74       2022-04-01   2022-07-01   0.022222 -0.003378  0.035   
75       2022-07-01   2022-10-01   0.024465 -0.001722  0.036   
76       2022-10-01   2023-01-01   0.022923 -0.000484  0.034   
77       2023-01-01   2023-04-01   0.026012  0.004668  0.034   

    actual_PIT_score         z  
0          -1.787565  0.496714  
1          -1.805684 -0.138264  
2          -1.844251  0.647689  
3          -1.882992  1.523030  
4          -1.903602 -0.234153  
..               ...       ...  
73         -1.974382  1.564644  
74         -2.009875 -2.619745  
75         -1.969204  0.821903  
76         -1.996815  0.087047  
77         -1.942942 -0.299007  

[78 rows x 7 columns]

In [34]:
# Fitted PIT SCORE:-
df['fitted_PIT_Score']=(threshold-np.sqrt(rho)*df['z'])/np.sqrt(1-rho)
df

observation_date Performance   Annual DR       HPI     UR  \
0        2003-10-01   2004-01-01   0.036923  0.010458  0.057   
1        2004-01-01   2004-04-01   0.035484  0.010575  0.056   
2        2004-04-01   2004-07-01   0.032573  0.009930  0.055   
3        2004-07-01   2004-10-01   0.029851  0.010189  0.055   
4        2004-10-01   2005-01-01   0.028481  0.012283  0.053   
..              ...          ...        ...       ...    ...   
73       2022-01-01   2022-04-01   0.024169  0.013562  0.037   
74       2022-04-01   2022-07-01   0.022222 -0.003378  0.035   
75       2022-07-01   2022-10-01   0.024465 -0.001722  0.036   
76       2022-10-01   2023-01-01   0.022923 -0.000484  0.034   
77       2023-01-01   2023-04-01   0.026012  0.004668  0.034   

    actual_PIT_score         z  fitted_PIT_Score  
0          -1.787565  0.496714         -1.680109  
1          -1.805684 -0.138264         -1.616291  
2          -1.844251  0.647689         -1.695282  
3          -1.882992  1.523030         -1.783257  
4          -1.903602 -0.234153         -1.606654  
..               ...       ...               ...  
73         -1.974382  1.564644         -1.787440  
74         -2.009875 -2.619745         -1.366893  
75         -1.969204  0.821903         -1.712791  
76         -1.996815  0.087047         -1.638936  
77         -1.942942 -0.299007         -1.600136  

[78 rows x 8 columns]

In [40]:
df['error']=(df['actual_PIT_score']-df['fitted_PIT_Score'])**2

In [100]:

df

observation_date Performance   Annual DR       HPI     UR  \
0        2003-10-01   2004-01-01   0.036923  0.010458  0.057   
1        2004-01-01   2004-04-01   0.035484  0.010575  0.056   
2        2004-04-01   2004-07-01   0.032573  0.009930  0.055   
3        2004-07-01   2004-10-01   0.029851  0.010189  0.055   
4        2004-10-01   2005-01-01   0.028481  0.012283  0.053   
..              ...          ...        ...       ...    ...   
73       2022-01-01   2022-04-01   0.024169  0.013562  0.037   
74       2022-04-01   2022-07-01   0.022222 -0.003378  0.035   
75       2022-07-01   2022-10-01   0.024465 -0.001722  0.036   
76       2022-10-01   2023-01-01   0.022923 -0.000484  0.034   
77       2023-01-01   2023-04-01   0.026012  0.004668  0.034   

    actual_PIT_score         z  fitted_PIT_Score     error  
0          -1.787565  0.496714         -1.680109  0.011547  
1          -1.805684 -0.138264         -1.616291  0.035870  
2          -1.844251  0.647689         -1.695282  0.022192  
3          -1.882992  1.523030         -1.783257  0.009947  
4          -1.903602 -0.234153         -1.606654  0.088178  
..               ...       ...               ...       ...  
73         -1.974382  1.564644         -1.787440  0.034947  
74         -2.009875 -2.619745         -1.366893  0.413426  
75         -1.969204  0.821903         -1.712791  0.065747  
76         -1.996815  0.087047         -1.638936  0.128078  
77         -1.942942 -0.299007         -1.600136  0.117516  

[78 rows x 9 columns]

In [52]:
initial_SSE=df['error'].sum()
initial_SSE

5.945732226391566

# Optimization using least square

1) Here we have to optimize Z values and rho.
2) For optimization using least squares we have to give the synatax as: least_square(function,initial parameters, bounds)

In [65]:
# Initial Parameters (give it as a list/array):-
initial_params=np.concatenate(([rho],df['z'].values))
initial_params

array([ 0.01      ,  0.49671415, -0.1382643 ,  0.64768854,  1.52302986,
       -0.23415337, -0.23413696,  1.57921282,  0.76743473, -0.46947439,
        0.54256004, -0.46341769, -0.46572975,  0.24196227, -1.91328024,
       -1.72491783, -0.56228753, -1.01283112,  0.31424733, -0.90802408,
       -1.4123037 ,  1.46564877, -0.2257763 ,  0.0675282 , -1.42474819,
       -0.54438272,  0.11092259, -1.15099358,  0.37569802, -0.60063869,
       -0.29169375, -0.60170661,  1.85227818, -0.01349722, -1.05771093,
        0.82254491, -1.22084365,  0.2088636 , -1.95967012, -1.32818605,
        0.19686124,  0.73846658,  0.17136828, -0.11564828, -0.3011037 ,
       -1.47852199, -0.71984421, -0.46063877,  1.05712223,  0.34361829,
       -1.76304016,  0.32408397, -0.38508228, -0.676922  ,  0.61167629,
        1.03099952,  0.93128012, -0.83921752, -0.30921238,  0.33126343,
        0.97554513, -0.47917424, -0.18565898, -1.10633497, -1.19620662,
        0.81252582,  1.35624003, -0.07201012,  1.0035329 ,  0.36

In [128]:
def objective(params):
    rho= params[0]
    z=params[1:]
    fitted_PIT_Score =(threshold-np.sqrt(rho)*z)/np.sqrt(1-rho)
    residuals= df['actual_PIT_score'] - fitted_PIT_Score
    return residuals

Constraints:-

1) z value range --> (-7 to +7)
2) rho >=0 (regulatory requirement)
3) variance of z = 1 (in least squares, it automatically makes the variance constant or 1.
4) The above constraints(bounds) can be created seperately or given while initializing, here we are giving seperately.

In [130]:
# Bounds LB for both z and rho--> (0,-7) UB for z and rho--> (1,+7)
bounds=([0]+[-7]*len(df),[1]+[7]*len(df))
bounds

([0,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7,
  -7],
 [1,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7,
  7])

In [132]:
from scipy.optimize import least_squares

In [134]:
results= least_squares(objective,initial_params,bounds=bounds)

In [135]:
optimal_rho=results.x[0]
optimal_rho

0.010764714291033678

In [147]:
# Optimal Z values in place of old z values:-
df['z']=results.x[1:]
df

observation_date Performance   Annual DR       HPI     UR  \
0        2003-10-01   2004-01-01   0.036923  0.010458  0.057   
1        2004-01-01   2004-04-01   0.035484  0.010575  0.056   
2        2004-04-01   2004-07-01   0.032573  0.009930  0.055   
3        2004-07-01   2004-10-01   0.029851  0.010189  0.055   
4        2004-10-01   2005-01-01   0.028481  0.012283  0.053   
..              ...          ...        ...       ...    ...   
73       2022-01-01   2022-04-01   0.024169  0.013562  0.037   
74       2022-04-01   2022-07-01   0.022222 -0.003378  0.035   
75       2022-07-01   2022-10-01   0.024465 -0.001722  0.036   
76       2022-10-01   2023-01-01   0.022923 -0.000484  0.034   
77       2023-01-01   2023-04-01   0.026012  0.004668  0.034   

    actual_PIT_score         z  fitted_PIT_Score     error  
0          -1.787565  1.502789         -1.680109  0.011547  
1          -1.805684  1.676488         -1.616291  0.035870  
2          -1.844251  2.046222         -1.695282  0.022192  
3          -1.882992  2.417620         -1.783257  0.009947  
4          -1.903602  2.615205         -1.606654  0.088178  
..               ...       ...               ...       ...  
73         -1.974382  3.293741         -1.787440  0.034947  
74         -2.009875  3.634225         -1.366893  0.413426  
75         -1.969204  3.244106         -1.712791  0.065747  
76         -1.996815  3.508802         -1.638936  0.128078  
77         -1.942942  2.992346         -1.600136  0.117516  

[78 rows x 9 columns]

In [149]:
# New Fitted_PIT_Score in place of Old Fitted_PIT_Score:-
df['fitted_PIT_Score']=(threshold-np.sqrt(optimal_rho)*df['z'])/np.sqrt(1-optimal_rho)
df

observation_date Performance   Annual DR       HPI     UR  \
0        2003-10-01   2004-01-01   0.036923  0.010458  0.057   
1        2004-01-01   2004-04-01   0.035484  0.010575  0.056   
2        2004-04-01   2004-07-01   0.032573  0.009930  0.055   
3        2004-07-01   2004-10-01   0.029851  0.010189  0.055   
4        2004-10-01   2005-01-01   0.028481  0.012283  0.053   
..              ...          ...        ...       ...    ...   
73       2022-01-01   2022-04-01   0.024169  0.013562  0.037   
74       2022-04-01   2022-07-01   0.022222 -0.003378  0.035   
75       2022-07-01   2022-10-01   0.024465 -0.001722  0.036   
76       2022-10-01   2023-01-01   0.022923 -0.000484  0.034   
77       2023-01-01   2023-04-01   0.026012  0.004668  0.034   

    actual_PIT_score         z  fitted_PIT_Score     error  
0          -1.787565  1.502789         -1.787582  0.011547  
1          -1.805684  1.676488         -1.805702  0.035870  
2          -1.844251  2.046222         -1.844271  0.022192  
3          -1.882992  2.417620         -1.883014  0.009947  
4          -1.903602  2.615205         -1.903625  0.088178  
..               ...       ...               ...       ...  
73         -1.974382  3.293741         -1.974407  0.034947  
74         -2.009875  3.634225         -2.009925  0.413426  
75         -1.969204  3.244106         -1.969230  0.065747  
76         -1.996815  3.508802         -1.996842  0.128078  
77         -1.942942  2.992346         -1.942967  0.117516  

[78 rows x 9 columns]

In [153]:
# New error:-
df['error']=(df['actual_PIT_score']- df['fitted_PIT_Score'])**2
df

observation_date Performance   Annual DR       HPI     UR  \
0        2003-10-01   2004-01-01   0.036923  0.010458  0.057   
1        2004-01-01   2004-04-01   0.035484  0.010575  0.056   
2        2004-04-01   2004-07-01   0.032573  0.009930  0.055   
3        2004-07-01   2004-10-01   0.029851  0.010189  0.055   
4        2004-10-01   2005-01-01   0.028481  0.012283  0.053   
..              ...          ...        ...       ...    ...   
73       2022-01-01   2022-04-01   0.024169  0.013562  0.037   
74       2022-04-01   2022-07-01   0.022222 -0.003378  0.035   
75       2022-07-01   2022-10-01   0.024465 -0.001722  0.036   
76       2022-10-01   2023-01-01   0.022923 -0.000484  0.034   
77       2023-01-01   2023-04-01   0.026012  0.004668  0.034   

    actual_PIT_score         z  fitted_PIT_Score         error  
0          -1.787565  1.502789         -1.787582  2.803632e-10  
1          -1.805684  1.676488         -1.805702  3.202404e-10  
2          -1.844251  2.046222         -1.844271  4.042765e-10  
3          -1.882992  2.417620         -1.883014  4.862210e-10  
4          -1.903602  2.615205         -1.903625  5.284201e-10  
..               ...       ...               ...           ...  
73         -1.974382  3.293741         -1.974407  6.631400e-10  
74         -2.009875  3.634225         -2.009925  2.548941e-09  
75         -1.969204  3.244106         -1.969230  6.539329e-10  
76         -1.996815  3.508802         -1.996842  7.068393e-10  
77         -1.942942  2.992346         -1.942967  6.044716e-10  

[78 rows x 9 columns]

In [157]:
final_SSE=df['error'].sum()
final_SSE

6.887211281749462e-07

We can aslo perform the above optimization using the minimize_scalar, we will have to give seperate code to make variance of z constant or 1.

# Regress z with the lagged values of macro eco variables

In [197]:
MEV_df=pd.DataFrame()

In [199]:
MEV_df['z']=df['z']


In [201]:
MEV_df['Lag_HPI']=df['HPI'].shift(1)
MEV_df['Lag_UR']=df['UR'].shift(1)
MEV_df

z   Lag_HPI  Lag_UR
0   1.502789       NaN     NaN
1   1.676488  0.010458   0.057
2   2.046222  0.010575   0.056
3   2.417620  0.009930   0.055
4   2.615205  0.010189   0.055
..       ...       ...     ...
73  3.293741  0.016720   0.040
74  3.634225  0.013562   0.037
75  3.244106 -0.003378   0.035
76  3.508802 -0.001722   0.036
77  2.992346 -0.000484   0.034

[78 rows x 3 columns]

In [203]:
# drop the 1st row, as it has null values:-
MEV_df.drop(0,axis=0,inplace=True)

In [205]:
MEV_df

z   Lag_HPI  Lag_UR
1   1.676488  0.010458   0.057
2   2.046222  0.010575   0.056
3   2.417620  0.009930   0.055
4   2.615205  0.010189   0.055
5   2.521173  0.012283   0.053
..       ...       ...     ...
73  3.293741  0.016720   0.040
74  3.634225  0.013562   0.037
75  3.244106 -0.003378   0.035
76  3.508802 -0.001722   0.036
77  2.992346 -0.000484   0.034

[77 rows x 3 columns]

In [209]:
# Regress Z with Lagged values of HPI and UR:-
from statsmodels.formula.api import ols
import statsmodels.api as sm

In [211]:
from statsmodels.stats.anova import anova_lm

In [219]:
model = ols('z ~ Lag_HPI + Lag_UR',data=MEV_df).fit()

In [223]:
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      z   R-squared:                       0.648
Model:                            OLS   Adj. R-squared:                  0.639
Method:                 Least Squares   F-statistic:                     68.23
Date:                Tue, 03 Feb 2026   Prob (F-statistic):           1.60e-17
Time:                        12:45:43   Log-Likelihood:                -141.79
No. Observations:                  77   AIC:                             289.6
Df Residuals:                      74   BIC:                             296.6
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      4.8290      0.550      8.777      0.000       3.733       5.925
Lag_HPI      136.1953     27.144      5.017      0.000      82.109     190.281
Lag_UR       -79.0243      8.238     -9.593      0.000     -95.438     -62.611
==============================================================================
Omnibus:                       75.942   Durbin-Watson:                   0.783
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              786.508
Skew:                           2.855   Prob(JB):                    1.63e-171
Kurtosis:                      17.578   Cond. No.                         154.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [227]:
anova_results=anova_lm(model)
anova_results

df      sum_sq     mean_sq          F        PR(>F)
Lag_HPI    1.0  107.657912  107.657912  44.438714  4.092368e-09
Lag_UR     1.0  222.947707  222.947707  92.027695  1.271742e-14
Residual  74.0  179.273537    2.422615        NaN           NaN

In [237]:
# We have the forecast of HPI and UR give by the macro eco analysis team:-

forecast_base=pd.read_excel('forecast.xlsx',sheet_name='Sheet1')
forecast_base

HPI      UR
0   0.0032  0.0581
1   0.0035  0.0592
2   0.0027  0.0564
3   0.0033  0.0587
4   0.0025  0.0559
5   0.0023  0.0553
6   0.0023  0.0553
7   0.0028  0.0570
8   0.0018  0.0536
9   0.0028  0.0570
10  0.0022  0.0548
11  0.0027  0.0564
12  0.0022  0.0548
13  0.0020  0.0542
14  0.0032  0.0581
15  0.0027  0.0564
16  0.0020  0.0542
17  0.0035  0.0592
18  0.0023  0.0553
19  0.0035  0.0592
20  0.0035  0.0592
21  0.0035  0.0592
22  0.0035  0.0592
23  0.0035  0.0592
24  0.0035  0.0592
25  0.0035  0.0592
26  0.0035  0.0592
27  0.0035  0.0592
28  0.0035  0.0592
29  0.0035  0.0592
30  0.0035  0.0592
31  0.0035  0.0592
32  0.0035  0.0592
33  0.0035  0.0592
34  0.0035  0.0592
35  0.0035  0.0592

In [247]:
forecast_base.rename(columns={'HPI' :'Lag_HPI', 'UR' : 'Lag_UR'},inplace=True)

In [249]:
forecast_base

Lag_HPI  Lag_UR
0    0.0032  0.0581
1    0.0035  0.0592
2    0.0027  0.0564
3    0.0033  0.0587
4    0.0025  0.0559
5    0.0023  0.0553
6    0.0023  0.0553
7    0.0028  0.0570
8    0.0018  0.0536
9    0.0028  0.0570
10   0.0022  0.0548
11   0.0027  0.0564
12   0.0022  0.0548
13   0.0020  0.0542
14   0.0032  0.0581
15   0.0027  0.0564
16   0.0020  0.0542
17   0.0035  0.0592
18   0.0023  0.0553
19   0.0035  0.0592
20   0.0035  0.0592
21   0.0035  0.0592
22   0.0035  0.0592
23   0.0035  0.0592
24   0.0035  0.0592
25   0.0035  0.0592
26   0.0035  0.0592
27   0.0035  0.0592
28   0.0035  0.0592
29   0.0035  0.0592
30   0.0035  0.0592
31   0.0035  0.0592
32   0.0035  0.0592
33   0.0035  0.0592
34   0.0035  0.0592
35   0.0035  0.0592

In [251]:
# Predict the Z values:-
MEF_pred=model.predict(forecast_base)
MEF_pred

0     0.673499
1     0.627431
2     0.739743
3     0.639704
4     0.752016
5     0.772191
6     0.772191
7     0.705948
8     0.838435
9     0.705948
10    0.798084
11    0.739743
12    0.798084
13    0.818260
14    0.673499
15    0.739743
16    0.818260
17    0.627431
18    0.772191
19    0.627431
20    0.627431
21    0.627431
22    0.627431
23    0.627431
24    0.627431
25    0.627431
26    0.627431
27    0.627431
28    0.627431
29    0.627431
30    0.627431
31    0.627431
32    0.627431
33    0.627431
34    0.627431
35    0.627431
dtype: float64

In [253]:
var_of_residual=anova_results.loc['Residual','mean_sq']
var_of_residual

2.4226153668574932

In [263]:
# Find the Forecasted PIT FL PD using vasicek:-
forecast_base['vasicek']=norm.cdf((threshold - np.sqrt(optimal_rho)*MEF_pred)/(np.sqrt(1-optimal_rho + optimal_rho * var_of_residual)))

In [265]:
forecast_base

Lag_HPI  Lag_UR   vasicek
0    0.0032  0.0581  0.046568
1    0.0035  0.0592  0.047032
2    0.0027  0.0564  0.045907
3    0.0033  0.0587  0.046908
4    0.0025  0.0559  0.045785
5    0.0023  0.0553  0.045586
6    0.0023  0.0553  0.045586
7    0.0028  0.0570  0.046243
8    0.0018  0.0536  0.044936
9    0.0028  0.0570  0.046243
10   0.0022  0.0548  0.045331
11   0.0027  0.0564  0.045907
12   0.0022  0.0548  0.045331
13   0.0020  0.0542  0.045133
14   0.0032  0.0581  0.046568
15   0.0027  0.0564  0.045907
16   0.0020  0.0542  0.045133
17   0.0035  0.0592  0.047032
18   0.0023  0.0553  0.045586
19   0.0035  0.0592  0.047032
20   0.0035  0.0592  0.047032
21   0.0035  0.0592  0.047032
22   0.0035  0.0592  0.047032
23   0.0035  0.0592  0.047032
24   0.0035  0.0592  0.047032
25   0.0035  0.0592  0.047032
26   0.0035  0.0592  0.047032
27   0.0035  0.0592  0.047032
28   0.0035  0.0592  0.047032
29   0.0035  0.0592  0.047032
30   0.0035  0.0592  0.047032
31   0.0035  0.0592  0.047032
32   0.0035  0.0592  0.047032
33   0.0035  0.0592  0.047032
34   0.0035  0.0592  0.047032
35   0.0035  0.0592  0.047032

Forecasted PIT FL PD for the next quarter is 0.046568.

# Now we calibrate using the log odds shift calibration:-


We have PIT FL PD for each snapshot, but we want PIT FL PD for each rating grade. TO adjest the PD curve based on target central tendency.

In [276]:
df1=pd.read_excel('1_TTC_PD.xlsx',sheet_name='TTC_PD')
df1

Snapshot       Date     SL  Default       id  bureau_score  num_ccj  \
0             1 2003-10-01  21912        0  4810802           308        0   
1             1 2003-10-01   4458        0  5047938           508        0   
2             1 2003-10-01  14523        0  8403557           496        0   
3             1 2003-10-01  13364        0  3140641           450        0   
4             1 2003-10-01   3733        1  8429363           366        0   
...         ...        ...    ...      ...      ...           ...      ...   
25456        78 2023-01-01  22041        0  5256024           285        1   
25457        78 2023-01-01   6311        0  5906396           484        0   
25458        78 2023-01-01  11770        0  6087526           460        0   
25459        78 2023-01-01  16455        0  4544378           392        0   
25460        78 2023-01-01  10576        0  7617583           520        0   

       max_arrears_12m  cc_util  annual_income  emp_length  \
0             0.185325   0.3995          56113           4   
1             0.000000   0.3005          49577           4   
2             0.000000   0.6321          66782          10   
3             0.000000   0.2897          44468           6   
4             2.796597   0.7797          22979           1   
...                ...      ...            ...         ...   
25456         0.000000   0.5625          59322           6   
25457         0.000000   0.5506          41811           2   
25458         0.000000   0.6717          57101           7   
25459         0.000000   0.5176          94068          11   
25460         0.000000   0.4811          40038           5   

       months_since_recent_cc_delinq   Pd_pred  Grade  
0                                 13  0.009485      9  
1                                 15  0.002070      7  
2                                  5  0.015489     10  
3                                 13  0.002667      7  
4                                  7  0.915764     17  
...                              ...       ...    ...  
25456                             13  0.024336     11  
25457                              9  0.026629     11  
25458                              4  0.039178     12  
25459                              3  0.003734      8  
25460                              8  0.012926     10  

[25461 rows x 14 columns]

In [283]:
TTC_PD=df1.groupby('Grade')[['Default','Pd_pred']].mean().reset_index()
TTC_PD

Grade   Default   Pd_pred
0       1  0.000000  0.000092
1       2  0.000000  0.000160
2       3  0.000000  0.000274
3       4  0.000000  0.000477
4       5  0.000000  0.000848
5       6  0.000000  0.001504
6       7  0.001111  0.002704
7       8  0.003949  0.004848
8       9  0.005373  0.008726
9      10  0.014212  0.015595
10     11  0.031130  0.027800
11     12  0.059238  0.049751
12     13  0.096141  0.090110
13     14  0.162879  0.161158
14     15  0.296024  0.293541
15     16  0.534351  0.535767
16     17  0.798077  0.843164

In [285]:
TTC_PD.rename(columns={'Default': 'ODR','Pd_pred': 'Model_PD'},inplace=True)
TTC_PD

Grade       ODR  Model_PD
0       1  0.000000  0.000092
1       2  0.000000  0.000160
2       3  0.000000  0.000274
3       4  0.000000  0.000477
4       5  0.000000  0.000848
5       6  0.000000  0.001504
6       7  0.001111  0.002704
7       8  0.003949  0.004848
8       9  0.005373  0.008726
9      10  0.014212  0.015595
10     11  0.031130  0.027800
11     12  0.059238  0.049751
12     13  0.096141  0.090110
13     14  0.162879  0.161158
14     15  0.296024  0.293541
15     16  0.534351  0.535767
16     17  0.798077  0.843164

In [287]:
TTC_PD['ODR']=TTC_PD['ODR'].apply(lambda x: 0.0003 if x==0 else x)
TTC_PD

Grade       ODR  Model_PD
0       1  0.000300  0.000092
1       2  0.000300  0.000160
2       3  0.000300  0.000274
3       4  0.000300  0.000477
4       5  0.000300  0.000848
5       6  0.000300  0.001504
6       7  0.001111  0.002704
7       8  0.003949  0.004848
8       9  0.005373  0.008726
9      10  0.014212  0.015595
10     11  0.031130  0.027800
11     12  0.059238  0.049751
12     13  0.096141  0.090110
13     14  0.162879  0.161158
14     15  0.296024  0.293541
15     16  0.534351  0.535767
16     17  0.798077  0.843164

In [307]:
distn= df1.groupby('Grade').size().reset_index(name='count')
distn

Grade  count
0       1      4
1       2     47
2       3    246
3       4    741
4       5   1529
5       6   2065
6       7   2701
7       8   3039
8       9   3164
9      10   3096
10     11   2602
11     12   2127
12     13   1529
13     14   1056
14     15    679
15     16    524
16     17    312

In [313]:
TTC_PD['%_Population']=distn['count']/distn['count'].sum()
TTC_PD

Grade       ODR  Model_PD  %_Population
0       1  0.000300  0.000092      0.000157
1       2  0.000300  0.000160      0.001846
2       3  0.000300  0.000274      0.009662
3       4  0.000300  0.000477      0.029103
4       5  0.000300  0.000848      0.060053
5       6  0.000300  0.001504      0.081104
6       7  0.001111  0.002704      0.106084
7       8  0.003949  0.004848      0.119359
8       9  0.005373  0.008726      0.124268
9      10  0.014212  0.015595      0.121598
10     11  0.031130  0.027800      0.102196
11     12  0.059238  0.049751      0.083540
12     13  0.096141  0.090110      0.060053
13     14  0.162879  0.161158      0.041475
14     15  0.296024  0.293541      0.026668
15     16  0.534351  0.535767      0.020580
16     17  0.798077  0.843164      0.012254

In [317]:
TTC_pd=(TTC_PD['ODR']*TTC_PD['%_Population']).sum()
TTC_pd

0.05236988335100742

In [319]:
temp=TTC_PD[['Grade','ODR','%_Population']]
temp

Grade       ODR  %_Population
0       1  0.000300      0.000157
1       2  0.000300      0.001846
2       3  0.000300      0.009662
3       4  0.000300      0.029103
4       5  0.000300      0.060053
5       6  0.000300      0.081104
6       7  0.001111      0.106084
7       8  0.003949      0.119359
8       9  0.005373      0.124268
9      10  0.014212      0.121598
10     11  0.031130      0.102196
11     12  0.059238      0.083540
12     13  0.096141      0.060053
13     14  0.162879      0.041475
14     15  0.296024      0.026668
15     16  0.534351      0.020580
16     17  0.798077      0.012254

In [327]:
temp['Log_Odds']=np.log(temp['ODR']/(1-temp['ODR']))
temp

Grade       ODR  %_Population  Log_Odds
0       1  0.000300      0.000157 -8.111428
1       2  0.000300      0.001846 -8.111428
2       3  0.000300      0.009662 -8.111428
3       4  0.000300      0.029103 -8.111428
4       5  0.000300      0.060053 -8.111428
5       6  0.000300      0.081104 -8.111428
6       7  0.001111      0.106084 -6.801654
7       8  0.003949      0.119359 -5.530421
8       9  0.005373      0.124268 -5.220992
9      10  0.014212      0.121598 -4.239363
10     11  0.031130      0.102196 -3.437962
11     12  0.059238      0.083540 -2.765120
12     13  0.096141      0.060053 -2.240854
13     14  0.162879      0.041475 -1.636963
14     15  0.296024      0.026668 -0.866306
15     16  0.534351      0.020580  0.137621
16     17  0.798077      0.012254  1.374318

In [331]:
from scipy.stats import linregress

In [335]:
slope,intercept,r,pvalue,std_error=linregress(x=temp['Grade'],y=temp['Log_Odds'])

In [337]:
slope

0.6299314590507241

In [339]:
intercept

-10.369163274316868

In [349]:
# intercept seed value:-
seed_intercept=-8

In [351]:
temp['fitted_log_odds']=seed_intercept + slope*temp['Grade']
temp

Grade       ODR  %_Population  Log_Odds  fitted_log_odds
0       1  0.000300      0.000157 -8.111428        -7.370069
1       2  0.000300      0.001846 -8.111428        -6.740137
2       3  0.000300      0.009662 -8.111428        -6.110206
3       4  0.000300      0.029103 -8.111428        -5.480274
4       5  0.000300      0.060053 -8.111428        -4.850343
5       6  0.000300      0.081104 -8.111428        -4.220411
6       7  0.001111      0.106084 -6.801654        -3.590480
7       8  0.003949      0.119359 -5.530421        -2.960548
8       9  0.005373      0.124268 -5.220992        -2.330617
9      10  0.014212      0.121598 -4.239363        -1.700685
10     11  0.031130      0.102196 -3.437962        -1.070754
11     12  0.059238      0.083540 -2.765120        -0.440822
12     13  0.096141      0.060053 -2.240854         0.189109
13     14  0.162879      0.041475 -1.636963         0.819040
14     15  0.296024      0.026668 -0.866306         1.448972
15     16  0.534351      0.020580  0.137621         2.078903
16     17  0.798077      0.012254  1.374318         2.708835

In [355]:
temp['Annual_Cal_PD']=np.exp(temp['fitted_log_odds'])/(1+np.exp(temp['fitted_log_odds']))
temp

Grade       ODR  %_Population  Log_Odds  fitted_log_odds  Annual_Cal_PD
0       1  0.000300      0.000157 -8.111428        -7.370069       0.000629
1       2  0.000300      0.001846 -8.111428        -6.740137       0.001181
2       3  0.000300      0.009662 -8.111428        -6.110206       0.002215
3       4  0.000300      0.029103 -8.111428        -5.480274       0.004151
4       5  0.000300      0.060053 -8.111428        -4.850343       0.007765
5       6  0.000300      0.081104 -8.111428        -4.220411       0.014480
6       7  0.001111      0.106084 -6.801654        -3.590480       0.026845
7       8  0.003949      0.119359 -5.530421        -2.960548       0.049240
8       9  0.005373      0.124268 -5.220992        -2.330617       0.088619
9      10  0.014212      0.121598 -4.239363        -1.700685       0.154376
10     11  0.031130      0.102196 -3.437962        -1.070754       0.255260
11     12  0.059238      0.083540 -2.765120        -0.440822       0.391545
12     13  0.096141      0.060053 -2.240854         0.189109       0.547137
13     14  0.162879      0.041475 -1.636963         0.819040       0.694033
14     15  0.296024      0.026668 -0.866306         1.448972       0.809840
15     16  0.534351      0.020580  0.137621         2.078903       0.888836
16     17  0.798077      0.012254  1.374318         2.708835       0.937546

In [357]:
TTC_PD_temp=(temp['Annual_Cal_PD']*temp['%_Population']).sum()
TTC_PD_temp

0.2121109389222412

In [390]:
tgt_pd=forecast_base['vasicek'][0]
tgt_pd

0.04656770743108183

In [361]:
from scipy.optimize import minimize_scalar

In [392]:
def obj_functn(adj_intercept):
    temp['fitted_log_odds']=adj_intercept + slope*temp['Grade']
    temp['Annual_Cal_PD']=np.exp(temp['fitted_log_odds'])/(1+np.exp(temp['fitted_log_odds']))
    TTC_PD_temp=(temp['Annual_Cal_PD']*temp['%_Population']).sum()
    return abs(TTC_PD_temp - tgt_pd)

In [394]:
result1=minimize_scalar(obj_functn)

In [396]:
result1.x

-10.437596841784915

In [398]:
result1.fun

6.237958968813295e-10

In [400]:
temp

Grade       ODR  %_Population  Log_Odds  fitted_log_odds  Annual_Cal_PD
0       1  0.000300      0.000157 -8.111428        -9.807666       0.000055
1       2  0.000300      0.001846 -8.111428        -9.177734       0.000103
2       3  0.000300      0.009662 -8.111428        -8.547803       0.000194
3       4  0.000300      0.029103 -8.111428        -7.917871       0.000364
4       5  0.000300      0.060053 -8.111428        -7.287940       0.000683
5       6  0.000300      0.081104 -8.111428        -6.658008       0.001282
6       7  0.001111      0.106084 -6.801654        -6.028077       0.002404
7       8  0.003949      0.119359 -5.530421        -5.398145       0.004505
8       9  0.005373      0.124268 -5.220992        -4.768214       0.008424
9      10  0.014212      0.121598 -4.239363        -4.138282       0.015700
10     11  0.031130      0.102196 -3.437962        -3.508351       0.029076
11     12  0.059238      0.083540 -2.765120        -2.878419       0.053231
12     13  0.096141      0.060053 -2.240854        -2.248488       0.095480
13     14  0.162879      0.041475 -1.636963        -1.618557       0.165404
14     15  0.296024      0.026668 -0.866306        -0.988625       0.271184
15     16  0.534351      0.020580  0.137621        -0.358694       0.411276
16     17  0.798077      0.012254  1.374318         0.271238       0.567397

In [402]:
TTC_PD_final=(temp['Annual_Cal_PD']*temp['%_Population']).sum()
TTC_PD_final

0.04656770137970711

Therefore We have found the PIT FL PD for each rating grade for the next Quarter.